03b_ipyautoui.ipynb

```diff
@@ -10,20 +10,749 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from typing_extensions import TypeAliasType"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ConstrainedInt = TypeAliasType(\"ConstrainedInt\", Annotated[int, Field(ge=1880, le=2020)])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class DataSelectorModel(BaseModel, validate_assignment=True):\n",
+    "    year_range: Annotated[\n",
+    "        # The key change is in the line below\n",
+    "        tuple[ConstrainedInt, ConstrainedInt],\n",
+    "        # With this change to the type we no longer need to tell ipyautoui\n",
+    "        # what kind of widget to use. Field contains just a brief description\n",
+    "        Field(description=\"Range of years to plot\")\n",
+    "    ] = (1800, 2000)\n",
+    "    window_size: Annotated[int, Field(ge=2, le=100)] = 2\n",
+    "    polynomial_order: Annotated[int, Field(ge=1, le=10)] = 1\n",
+    "\n",
+    "    # mode=\"after\" means the validator runs after pydantic has checked that the individual\n",
+    "    # fields have values that are valid.\n",
+    "    @model_validator(mode=\"after\")\n",
+    "    def limit_polynomial_order(self):\n",
+    "        \n",
+    "        if self.polynomial_order > self.window_size - 1:\n",
+    "            # Handle a bad polynomial order or window size\n",
+    "            raise ValueError(\"Polynomial order must be smaller than window size\")\n",
+    "            \n",
+    "        # If we got this far the polynomial order is consistent with the window size\n",
+    "        # so return self. Failing to return self will end up causing an error.\n",
+    "        return self"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ui_final = AutoUi(DataSelectorModel)\n",
+    "ui_final"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Enforcing the constraint between polynomial order and window size\n",
+    "\n",
+    "There are a few options here:\n",
+    "\n",
+    "1. Change one or both of the ranges of `widnow_size` and `polynomial_order`\n",
+    "    1. When window size is being changed and a conflict arises, either\n",
+    "        1. update the maximum value of the polynomial order to be consistent with window size, with the side effect that the value will be updated to.\n",
+    "        1. update the value of `polynomial_order` but not its maximum.\n",
+    "    1. When the polynomial order is changed, either\n",
+    "        1. update the minimum allowed window size, or\n",
+    "        1. update the value of the window size.\n",
+    "1. reset the controls to a valid value and display a warning to the user.\n",
+    "\n",
```

Comment:
I would suggest using variable names for all of these rather than the mixture of `polynomial_order`/`window_size` and "polynomial order" and "window size".